network-config-analyzer/IstioSidecar.py

```diff
@@ -0,0 +1,137 @@
+
+
+@dataclass
+class IstioSidecarRule:
+    """
+    A class representing a single egress rule (IstioEgressListener) in an Istio Sidecar object
+    """
+
+    def __init__(self, peer_set, peers_for_ns_compare):
+        """
+        Init the Egress rule of an Istio Sidecar
+        :param Peer.PeerSet peer_set: The set of mesh internal peers this rule allows connection to
+        :param Peer.PeerSet peers_for_ns_compare: The set of peers captured by a global sidecar with hosts
+        having namespace equal to '.'
+        """
+        self.egress_peer_set = peer_set
+        self.special_egress_peer_set = peers_for_ns_compare  # set of peers captured by a global sidecar with hosts of
+        # './<any>' form - then peers in this set will be in allowed connections
+        # only if are in the same namespace of the source peer captured by the sidecar
+
+
+class IstioSidecar(NetworkPolicy):
+    """
+    This class implements istio-specific logic for Sidecar
+    """
+
+    def __init__(self, name, namespace):
+        super().__init__(name, namespace)
+        self.default_sidecar = False  # a flag that indicates if the sidecar is selector-less (default) or not
+
+    def __eq__(self, other):
+        return super().__eq__(other) and self.default_sidecar == other.default_sidecar
+
+    def allowed_connections(self, from_peer, to_peer, is_ingress):
+        """
+        Evaluate the set of connections this policy allows/denies/passes between two peers
+        :param Peer.Peer from_peer: The source peer
+        :param Peer.Peer to_peer:  The target peer
+        :param bool is_ingress: whether we evaluate ingress rules only or egress rules only
+        :return: A PolicyConnections object containing sets of allowed/denied/pass connections
+        :rtype: PolicyConnections
+        """
+        # currently not handling ingress
+        if is_ingress:
+            return PolicyConnections(False, ConnectionSet(True))
+
+        captured = from_peer in self.selected_peers
+        # if not captured, or captured but the sidecar is not in from_peer top priority, don't consider connections
+        if not captured or (captured and not self._is_sidecar_prior(from_peer)):
+            return PolicyConnections(False)
+
+        conns = ConnectionSet(True)
+        # since sidecar rules include only peer sets for now, if a to_peer appears in any rule then connections allowed
+        for rule in self.egress_rules:
+            if to_peer in rule.egress_peer_set or \
+                    (to_peer in rule.special_egress_peer_set and from_peer.namespace == to_peer.namespace):
+                return PolicyConnections(True, allowed_conns=conns)
+
+        # egress from from_peer to to_peer is not allowed : if to_peer not been captured in the rules' egress_peer_set,
+        # or if the sidecar is global and to_peer is not in same namespace of from_peer while rule host's ns is '.'
+        return PolicyConnections(True, denied_conns=conns)
```

Comment:
Shouldn't return here `PolicyConnections(True, allowed_conns=ConnectionSet())` ? (there is no 'deny' action for sidecar, so I think this should be considered as empty allowed connections rather than denied connections of All connections).

Reply:
True, fixing this